src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,655 @@
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (writer is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.writer);
+            }
+
+            // Delegate to the Span-based method if possible.
+            // If that doesn't work, allocate the Encoder instance and run a loop.
+
+            if (bytes.IsSingleSegment)
+            {
+                return GetChars(encoding!, bytes.FirstSpan, writer!);
+            }
+            else
+            {
+                Convert(encoding!.GetDecoder(), bytes, writer!, flush: true, out long charsWritten, out bool completed);
+                Debug.Assert(completed, "Should be no intermediate state left over in the Decoder.");
+
+                return charsWritten;
+            }
+        }
+
+        /// <summary>
+        /// Decodes the specified <see cref="ReadOnlySequence{Byte}"/> to <see langword="char"/>s using the specified <see cref="Encoding"/>
+        /// and outputs the result to <paramref name="chars"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="bytes"/> is encoded.</param>
+        /// <param name="bytes">The <see cref="ReadOnlySequence{Byte}"/> to decode to characters.</param>
+        /// <param name="chars">The destination buffer to which the decoded characters will be written.</param>
+        /// <returns>The number of chars written to <paramref name="chars"/>.</returns>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="chars"/> is not large enough to contain the encoded form of <paramref name="bytes"/>.</exception>
+        /// <exception cref="DecoderFallbackException">Thrown if <paramref name="bytes"/> contains data that cannot be decoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static int GetChars(this Encoding encoding, in ReadOnlySequence<byte> bytes, Span<char> chars)
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (bytes.IsSingleSegment)
+            {
+                // If the incoming sequence is single-segment, one-shot this.
+
+                return encoding!.GetChars(bytes.FirstSpan, chars);
+            }
+            else
+            {
+                // If the incoming sequence is multi-segment, create a stateful Decoder
+                // and use it as the workhorse. On the final iteration we'll pass flush=true.
+
+                ReadOnlySequence<byte> remainingBytes = bytes;
+                int originalCharsLength = chars.Length;
```

Comment:
Why is this local copy necessary?  I don't see chars getting reassigned anywhere.